src/v/redpanda/admin/field_registry.cc

```diff
@@ -0,0 +1,148 @@
+  const std::string& field_path, const std::vector<int32_t>& field_numbers) {
+    if constexpr (std::is_same_v<ValueType, std::monostate>) {
+        throw std::invalid_argument(
+          "Cannot create accessor for unset field: " + field_path);
+    } else if constexpr (requires {
+                             typename field_getter_type<T, ValueType>::type;
+                         }) {
+        using GetterType = typename field_getter_type<T, ValueType>::type;
+        using ReturnType = std::invoke_result_t<typename GetterType::type, T&>;
+
+        return field_accessor_info<T>{
+          .getter = GetterType{[field_numbers](T& obj) {
+              return extract_field_value<ReturnType>(obj, field_numbers);
+          }}};
+    } else {
+        throw std::invalid_argument(
+          "Unsupported field type for filtering: " + field_path);
+    }
+}
+
+template<RedpandaProtobufMessage T>
+template<typename ReturnType>
+ReturnType protobuf_field_registry<T>::extract_field_value(
+  T& obj, const std::vector<int32_t>& field_numbers) {
+    auto field_opt = obj.lookup_field(field_numbers);
+    if (!field_opt) {
+        throw std::runtime_error("Field lookup failed during extraction");
+    }
+
+    return std::visit(
+      [](const auto& value) -> ReturnType {
+          using ValueType = std::decay_t<decltype(value)>;
+
+          if constexpr (std::is_same_v<ReturnType, ValueType>) {
+              return value;
+          } else if constexpr (std::is_same_v<ReturnType, int64_t>) {
+              if constexpr (std::is_same_v<ValueType, int32_t>) {
+                  return static_cast<int64_t>(value);
+              }
+          } else if constexpr (std::is_same_v<ReturnType, uint64_t>) {
+              if constexpr (std::is_same_v<ValueType, uint32_t>) {
+                  return static_cast<uint64_t>(value);
+              }
+          } else if constexpr (std::is_same_v<ReturnType, double>) {
+              if constexpr (std::is_same_v<ValueType, float>) {
+                  return static_cast<double>(value);
+              }
+          } else if constexpr (std::is_same_v<ReturnType, ss::sstring>) {
+              if constexpr (std::
+                              is_same_v<ValueType, serde::pb::raw_enum_value>) {
+                  return ss::sstring{value.name};
+              }
+          }
+          throw std::runtime_error("Unsupported field value type conversion");
+      },
+      field_opt->value);
+}
+
+template<typename T>
+std::unique_ptr<field_registry<T>> make_field_registry() {
+    return std::make_unique<protobuf_field_registry<T>>();
+}
+
+// Explicitly instantiating for all supported protobuf types to avoid having to
+// expose all the templated code in the header
+template class protobuf_field_registry<aip_filter_test::test_message>;
+template std::unique_ptr<field_registry<aip_filter_test::test_message>>
+make_field_registry<aip_filter_test::test_message>();
```

Comment:
this feels a bit strange pattern. Shouldn't instead we do something like replace `make_field_registry` in the header with `make_aip_filter_test_field_registry`?